### PR TITLE
Fix: No replicas for donors in HCA (#6582)

### LINKED ISSUE #6582: No replicas for donors in HCA

### LINKED ISSUE #6596: FIXMEs reference closed external issues
https://github.com/DataBiosphere/azul/blob/33796d4de83ff10fab3081ec28b3b0e2a2a2f18e/src/azul/plugins/metadata/hca/indexer/transform.py#L1392

[The linked issue](https://github.com/HumanCellAtlas/metadata-schema/issues/623) has been closed. This may manifest as a bug where zarr files are missing from replicas' lists of hub IDs and from the inner entity arrays of cell suspensions and samples.

### PR DESCRIPTION
**Pull request by nadove-ucsc**
Connected issues: #6582

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] On ZenHub, PR is connected to all issues it (partially) resolves
- [x] PR description links to connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] PR title references all connected issues
- [x] For each connected issue, there is at least one commit whose title references that issue

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (partiality)

- [x] Added `p` tag to titles of partial commits
- [x] This PR is labeled `partial` <sub>or completely resolves all connected issues</sub>
- [x] This PR partially resolves each of the connected issues <sub>or does not have the `partial` label</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or is not chained to another PR</sub>
- [x] The blocking PR is labeled `base` <sub>or this PR is not chained to another PR</sub>
- [x] This PR is labeled `chained` <sub>or is not chained to another PR</sub>

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or the changes introduced by this PR will not require reindexing of any deployment</sub>
- [x] This PR is labeled `reindex:dev` <sub>or the changes introduced by it will not require reindexing of `dev`</sub>
- [x] This PR is labeled `reindex:anvildev` <sub>or the changes introduced by it will not require reindexing of `anvildev`</sub>
- [x] This PR is labeled `reindex:anvilprod` <sub>or the changes introduced by it will not require reindexing of `anvilprod`</sub>
- [x] This PR is labeled `reindex:prod` <sub>or the changes introduced by it will not require reindexing of `prod`</sub>
- [x] This PR is labeled `reindex:partial` and its description documents the specific reindexing procedure for `dev`, `anvildev`, `anvilprod` and `prod` <sub>or requires a full reindex or carries none of the labels `reindex:dev`, `reindex:anvildev`, `reindex:anvilprod` and `reindex:prod`</sub>
- [x] This PR and its connected issues are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>

### Author (upgrading deployments)

- [x] Ran `make docker_images.json` and committed the resulting changes <sub>or this PR does not modify `azul_docker_images`, or any other variables referenced in the definition of that variable</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading deployments</sub>
- [x] This PR is labeled `upgrade` <sub>or does not require upgrading deployments</sub>
- [x] This PR is labeled `deploy:shared` <sub>or does not modify `docker_images.json`, and does not require deploying the `shared` component for any other reason</sub>
- [x] This PR is labeled `deploy:gitlab` <sub>or does not require deploying the `gitlab` component</sub>
- [x] This PR is labeled `deploy:runner` <sub>or does not require deploying the `runner` image</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the none of the stable branches (`anvilprod` and `prod`) have temporary hotfixes for any of the issues connected to this PR</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not modify `requirements*.txt`, `common.mk`, `Makefile` and `Dockerfile`</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not modify `requirements*.txt`</sub>
- [x] This PR is labeled `reqs` <sub>or does not modify `requirements*.txt`</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not modify functionality that could affect the IT outcome</sub>

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] PR is awaiting requested review from system administrator
- [x] PR is assigned to only the system administrator

### System administrator (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] A comment to this PR details the completed security design review
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved connected issues to *Approved* column
- [x] PR is assigned to only the operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex:…` labels and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Ran `_select dev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select dev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Ran `_select anvildev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Checked the items in the next section <sub>or this PR is labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the system administrator <sub>or this PR is not labeled `deploy:gitlab`</sub>

### System administrator

- [x] Background migrations for `dev.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Background migrations for `anvildev.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the operator

### Operator (before pushing merge the commit)

- [x] Ran `_select dev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Ran `_select anvildev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `dev`</sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvildev`</sub>
- [x] Started reindex in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] The title of the merge commit starts with the title of this PR
- [x] Added PR # reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but only included `p` if the PR is also labeled `partial`</sub>
- [x] Moved connected issues to *Merged lower* column in ZenHub
- [x] Moved blocked issues to *Triage* <sub>or no issues are blocked on the connected issues</sub>
- [x] Pushed merge commit to GitHub

### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>

### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev`
- [x] Pushed merge commit to GitLab `anvildev`
- [x] Build passes on GitLab `dev`
- [x] Reviewed build logs for anomalies on GitLab `dev`
- [x] Build passes on GitLab `anvildev`
- [x] Reviewed build logs for anomalies on GitLab `anvildev`
- [x] Ran `_select dev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

### Operator (reindex)

- [x] Deindexed all unreferenced catalogs in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed all unreferenced catalogs in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Deindexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Indexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Indexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Emptied fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Emptied fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>

### Operator

- [x] Propagated the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels to the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] Propagated any specific instructions related to the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels, from the description of this PR to that of the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] PR is assigned to no one

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: src/humancellatlas/data/metadata/api.py [hunk 1]
```diff
@@ -202,9 +203,8 @@ def accept(self, visitor: EntityVisitor):
 class LinkError(RuntimeError):
 
     def __init__(self, entity: LinkedEntity, other_entity: Entity, forward: bool) -> None:
-        super().__init__(entity.address +
-                         ' cannot ' + ('reference ' if forward else 'be referenced by ') +
-                         other_entity.address)
+        super().__init__(f'{entity.ref} cannot {"reference" if forward else "be referenced by"} '
```
Comment (achave11-ucsc) [inline src/humancellatlas/data/metadata/api.py]: L

File: src/azul/indexer/transform.py [hunk 2]
```diff
@@ -87,7 +85,7 @@ def estimate(self, partition: BundlePartition) -> int:
         """
 
     @abstractmethod
-    def transform(self, partition: BundlePartition) -> Iterable[Transform]:
+    def transform(self, partition: BundlePartition) -> Iterable[Contribution | Replica]:
```
Comment (achave11-ucsc) [inline src/azul/indexer/transform.py]: L

File: src/azul/plugins/metadata/anvil/indexer/transform.py [hunk 3]
```diff
@@ -168,32 +165,20 @@ def aggregator(cls, entity_type) -> EntityAggregator:
     def estimate(self, partition: BundlePartition) -> int:
         return sum(map(partial(self._contains, partition), self.bundle.entities))
 
-    def transform(self, partition: BundlePartition) -> Iterable[Transform]:
-        return (
-            self._transform(entity)
-            for entity in self._list_entities()
-            if self._contains(partition, entity)
-        )
+    def transform(self, partition: BundlePartition) -> Iterable[Contribution | Replica]:
```
Comment (achave11-ucsc) [inline src/azul/plugins/metadata/anvil/indexer/transform.py]: L

File: src/azul/plugins/metadata/hca/indexer/transform.py [hunk 4]
```diff
@@ -1418,7 +1406,7 @@ def _entities_in(self, partition: BundlePartition) -> Iterator[ENTITY]:
     def estimate(self, partition: BundlePartition) -> int:
         return ilen(self._entities_in(partition))
 
-    def transform(self, partition: BundlePartition) -> Iterable[Transform]:
+    def transform(self, partition: BundlePartition) -> Iterable[Contribution | Replica]:
```
Comment (achave11-ucsc) [inline src/azul/plugins/metadata/hca/indexer/transform.py]: L

File: src/azul/plugins/metadata/anvil/indexer/transform.py [hunk 5]
```diff
@@ -619,7 +597,7 @@ class DonorTransformer(BaseTransformer):
     def entity_type(cls) -> str:
         return 'donors'
 
-    def _transform(self, entity: EntityReference) -> Transform:
+    def _transform(self, entity: EntityReference) -> Iterable[Contribution | Replica]:
```
Comment (achave11-ucsc) [inline src/azul/plugins/metadata/anvil/indexer/transform.py]: L

File: src/azul/plugins/metadata/anvil/indexer/transform.py [hunk 6]
```diff
@@ -643,7 +623,7 @@ class FileTransformer(BaseTransformer):
     def entity_type(cls) -> str:
         return 'files'
 
-    def _transform(self, entity: EntityReference) -> Transform:
+    def _transform(self, entity: EntityReference) -> Iterable[Contribution | Replica]:
```
Comment (achave11-ucsc) [inline src/azul/plugins/metadata/anvil/indexer/transform.py]: L

File: src/azul/plugins/metadata/anvil/indexer/transform.py [hunk 7]
```diff
@@ -168,32 +165,20 @@ def aggregator(cls, entity_type) -> EntityAggregator:
...
+        for entity in self._list_entities():
+            if self._contains(partition, entity):
+                yield from self._transform(entity)
 
     def _list_entities(self) -> Iterable[EntityReference]:
         return self.bundle.entities
 
     @abstractmethod
-    def _transform(self, entity: EntityReference) -> Transform:
+    def _transform(self, entity: EntityReference) -> Iterable[Contribution | Replica]:
```
Comment (achave11-ucsc) [inline src/azul/plugins/metadata/anvil/indexer/transform.py]: L

File: src/azul/plugins/metadata/anvil/indexer/transform.py [hunk 8]
```diff
@@ -524,7 +496,7 @@ class ActivityTransformer(BaseTransformer):
     def entity_type(cls) -> str:
         return 'activities'
 
-    def _transform(self, entity: EntityReference) -> Transform:
+    def _transform(self, entity: EntityReference) -> Iterable[Contribution | Replica]:
```
Comment (achave11-ucsc) [inline src/azul/plugins/metadata/anvil/indexer/transform.py]: def _transform(self, entity: EntityReference) -> Iterable[Contribution | Replica]:

File: src/azul/plugins/metadata/anvil/indexer/transform.py [hunk 9]
```diff
@@ -545,7 +519,7 @@ class BiosampleTransformer(BaseTransformer):
     def entity_type(cls) -> str:
         return 'biosamples'
 
-    def _transform(self, entity: EntityReference) -> Transform:
+    def _transform(self, entity: EntityReference) -> Iterable[Contribution | Replica]:
```
Comment (achave11-ucsc) [inline src/azul/plugins/metadata/anvil/indexer/transform.py]: L

File: src/azul/plugins/metadata/hca/indexer/transform.py [hunk 10]
```diff
@@ -1431,7 +1419,7 @@ def entity_type(cls) -> str:
     def _entities(self) -> Iterable[api.File]:
         return self.api_bundle.not_stitched(self.api_bundle.files)
 
-    def _transform(self, files: Iterable[api.File]) -> Iterable[Contribution]:
+    def _transform(self, files: Iterable[api.File]) -> Iterable[Contribution | Replica]:
```
Comment (achave11-ucsc) [inline src/azul/plugins/metadata/hca/indexer/transform.py]: L

File: src/azul/plugins/metadata/hca/indexer/transform.py [hunk 11]
```diff
@@ -1649,11 +1643,11 @@ def _dated_entities(self) -> Iterable[DatedEntity]:
     def estimate(self, partition: BundlePartition) -> int:
         return int(partition.contains(self._singleton_id))
 
-    def transform(self, partition: BundlePartition) -> Iterable[Transform]:
+    def transform(self, partition: BundlePartition) -> Iterable[Contribution | Replica]:
```
Comment (achave11-ucsc) [inline src/azul/plugins/metadata/hca/indexer/transform.py]: L

File: src/azul/plugins/metadata/hca/indexer/transform.py [hunk 12]
```diff
@@ -1458,10 +1465,20 @@ def _transform(self, files: Iterable[api.File]) -> Iterable[Contribution | Repli
...
-                    hub_ids = list(map(str, visitor.files))
-                    yield self._replica(file.ref, hub_ids)
+                    yield self._replica(self.api_bundle.ref,
+                                        # The other hubs will be added when the indexer consolidates duplicate replicas.
+                                        [file_id])
+                    yield self._replica(one(self.api_bundle.projects.values()).ref,
+                                        # Projects are linked to every file in their snapshot, making an explicit list
+                                        # of hub IDs for the project both redundant and impractically large. Therefore,
+                                        # we leave the hub IDs field empty for projects and rely on the tenet that every
+                                        # file is an implicit hub of its parent project.
```
Comment (achave11-ucsc) [inline src/azul/plugins/metadata/hca/indexer/transform.py]: L, https://github.com/DataBiosphere/azul/blob/develop/CONTRIBUTING.rst#12line-length
```diff
diff --git a/src/azul/plugins/metadata/hca/indexer/transform.py b/src/azul/plugins/metadata/hca/indexer/transform.py
index d263e7c2..f1942292 100644
--- a/src/azul/plugins/metadata/hca/indexer/transform.py
+++ b/src/azul/plugins/metadata/hca/indexer/transform.py
@@ -1469,13 +1469,19 @@ class FileTransformer(PartitionedTransformer[api.File]):
                 yield self._contribution(contents, file_id)
                 if config.enable_replicas:
                     yield self._replica(self.api_bundle.ref,
-                                        # The other hubs will be added when the indexer consolidates duplicate replicas.
+                                        # The other hubs will be added when the
+                                        # indexer consolidates duplicate
+                                        # replicas.
                                         [file_id])
                     yield self._replica(one(self.api_bundle.projects.values()).ref,
-                                        # Projects are linked to every file in their snapshot, making an explicit list
-                                        # of hub IDs for the project both redundant and impractically large. Therefore,
-                                        # we leave the hub IDs field empty for projects and rely on the tenet that every
-                                        # file is an implicit hub of its parent project.
+                                        # Projects are linked to every file in
+                                        # their snapshot, making an explicit
+                                        # list of hub IDs for the project both
+                                        # redundant and impractically large.
+                                        # Therefore, we leave the hub IDs field
+                                        # empty for projects and rely on the
+                                        # tenet that every file is an implicit
+                                        # hub of its parent project.
                                         [])
                     for linked_entity in visitor.entities:
                         yield self._replica(linked_entity, [file_id])
```

File: src/azul/plugins/metadata/anvil/indexer/transform.py [hunk 13]
```diff
@@ -656,8 +608,14 @@ def _transform(self, entity: EntityReference) -> Transform:
...
-        return self._add_replica(contents, entity, hub_ids)
+        yield self._contribution(contents, entity.entity_id)
+        if config.enable_replicas:
+            # The other hubs will be added when the indexer consolidates duplicate replicas.
+            yield self._replica(entity, [entity.entity_id])
+            for linked_entity in linked:
+                # Datasets are linked to every file in their snapshot, making an explicit list
+                # of hub IDs for the dataset both redundant and impractically large. Therefore,
+                # we leave the hub IDs field empty for datasets and rely on the tenet that every
+                # file is an implicit hub of its parent dataset.
```
Comment (achave11-ucsc) [inline src/azul/plugins/metadata/anvil/indexer/transform.py]: L
```diff
diff --git a/src/azul/plugins/metadata/anvil/indexer/transform.py b/src/azul/plugins/metadata/anvil/indexer/transform.py
index 06464c9a..5e004beb 100644
--- a/src/azul/plugins/metadata/anvil/indexer/transform.py
+++ b/src/azul/plugins/metadata/anvil/indexer/transform.py
@@ -610,12 +610,14 @@ class FileTransformer(BaseTransformer):
         )
         yield self._contribution(contents, entity.entity_id)
         if config.enable_replicas:
-            # The other hubs will be added when the indexer consolidates duplicate replicas.
+            # The other hubs will be added when the indexer consolidates
+            # duplicate replicas.
             yield self._replica(entity, [entity.entity_id])
             for linked_entity in linked:
-                # Datasets are linked to every file in their snapshot, making an explicit list
-                # of hub IDs for the dataset both redundant and impractically large. Therefore,
-                # we leave the hub IDs field empty for datasets and rely on the tenet that every
-                # file is an implicit hub of its parent dataset.
+                # Datasets are linked to every file in their snapshot, making an
+                # explicit list of hub IDs for the dataset both redundant and
+                # impractically large. Therefore, we leave the hub IDs field
+                # empty for datasets and rely on the tenet that every file is an
+                # implicit hub of its parent dataset.
                 yield self._replica(linked_entity,
                                     hub_ids=[] if linked_entity.entity_type == 'dataset' else [entity.entity_id])

```

File: src/azul/plugins/metadata/hca/indexer/transform.py [hunk 14]
```diff
@@ -1393,14 +1380,21 @@ def visit(self, entity: api.Entity) -> None:
...
+
 
 ENTITY = TypeVar('ENTITY', bound=api.Entity)
 
 
 class PartitionedTransformer(BaseTransformer, Generic[ENTITY]):
 
     @abstractmethod
-    def _transform(self, entities: Iterable[ENTITY]) -> Iterable[Transform]:
+    def _transform(self, entities: Iterable[ENTITY]) -> Iterable[Contribution | Replica]:
```
Comment (achave11-ucsc) [inline src/azul/plugins/metadata/hca/indexer/transform.py]: L

**Review by achave11-ucsc — CHANGES_REQUESTED** [review]
Looks good.

Just spotted some minor nits to fix.

**Comment by nadove-ucsc** [comment]
Rebasing to address conflicts.

**Review by achave11-ucsc — DISMISSED** [review]
Approved ✅

File: test/indexer/test_indexer.py [hunk 15]
```diff
@@ -166,19 +150,16 @@ def _assert_hit_counts(self,
         :param num_aggs: Expected number of aggregates. If unspecified,
                          ``num_contribs`` becomes the default.
 
-        :param num_replicas: Expected number of replicas. If unspecified,
-                             ``num_contribs`` is used to calculate it, assuming
-                             all contributions have distinct contents.
+        :param num_replicas: Expected number of replicas, assuming replicas are
```
Comment (dsotirho-ucsc) [inline test/indexer/test_indexer.py:150]: Should be moved above `num_aggs`

**Review by dsotirho-ucsc — CHANGES_REQUESTED** [review]
Nicely done. Just one documentation nit.

**Review by dsotirho-ucsc — DISMISSED** [review]
Approved.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
A few nits, see REVIEW commit.

**Review by dsotirho-ucsc — APPROVED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
I added the `reindex:anvildev` and `reindex:anvilprod` labels because I would like to verify the impact of these changes for both AnVIL and HCA. As in the previous PR on this track, I would like to run a reindex immediately before and after merging the PR, and compare the runs via the `dev` and `anvildev` CW dashboards.

**Comment by hannes-ucsc** [comment]
### Security design review

- [x] Security design review completed; this PR does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with Terra 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

**Comment by achave11-ucsc** [comment]
`dev`:
Prior reindex: https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/101494
Posterior reindex: https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/101674

`anvildev`:
Prior reindex: https://gitlab.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/49234
Posterior reindex: https://gitlab.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/49426

**Pull request merged**